### PR TITLE
casr-afl

### PR DESCRIPTION
**Pull request by anfedotoff**

### CODE REVIEW

File: src/bin/casr-afl.rs [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    // Init log.
+    let log_level = if matches.value_of("log-level").unwrap() == "debug" {
+        LevelFilter::Debug
+    } else {
+        LevelFilter::Info
+    };
+    let _ = TermLogger::init(
+        log_level,
+        ConfigBuilder::new().set_time_to_local(true).build(),
+        TerminalMode::Mixed,
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:89]: Maybe Stderr?

File: src/bin/casr-afl.rs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+        // Get crashes from one node.
+        let mut crash_info = AflCrashInfo::default();
+        let cmdline_path = path.join("cmdline");
+        if let Ok(cmdline) = fs::read_to_string(&cmdline_path) {
+            crash_info.target_args = cmdline.split_whitespace().map(|s| s.to_string()).collect();
+            crash_info.is_stdin = !cmdline.contains("@@");
+
+            // Check for ASAN symbols.
+            if let Some(target) = crash_info.target_args.first() {
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:108]: Maybe we should create function and place it to util?

File: src/bin/casr-afl.rs [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+                .starts_with("id:")
+            {
+                crash_info.path = crash_path.to_path_buf();
+                crashes.push(crash_info.clone());
+            }
+        }
+    }
+
+    // Generate CASR reports.
+    info!("Generating CASR reports...");
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:153]: Log number of crashes

File: src/bin/casr-afl.rs [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        fs::create_dir(output_dir).with_context(|| {
+            format!("Couldn't create output directory {}", output_dir.display())
+        })?;
+    }
+
+    for crash in crashes {
+        let mut args: Vec<String> = vec!["-o".to_string()];
+        if crash.is_asan {
+            args.push(format!(
+                "{}.casrep",
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:165]: you may just create different format strings, the left code is similar (repeating)

File: src/bin/casr-afl.rs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        }
+        if crash.is_stdin {
+            args.push("--stdin".to_string());
+            args.push(crash.path.to_str().unwrap().to_string());
+            args.push("--".to_string());
+            args.extend_from_slice(&crash.target_args);
+        } else {
+            args.push("--".to_string());
+            args.extend_from_slice(&crash.target_args);
+            if let Some(at_index) = args.iter().position(|s| s.contains("@@")) {
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:182]: You should iter only in target_args, because crash name may contain @@

Comment (SweetVishnya) [reply src/bin/casr-afl.rs:182]: You may just save @@ index while creating CrashInfo structures

File: src/bin/casr-afl.rs [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+                let input = args[at_index].replace("@@", crash.path.to_str().unwrap());
+                args[at_index] = input;
+            }
+        }
+        let tool = if crash.is_asan {
+            "casr-san"
+        } else {
+            "casr-gdb"
+        };
+        debug!("{} {}", tool, args.join(" "));
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:192]: You'd better print Command via "{:?}"

Comment (SweetVishnya) [reply src/bin/casr-afl.rs:192]: join " " does not mind spaces

File: src/bin/casr-afl.rs [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        let casr_cmd = Command::new(tool)
+            .args(&args)
+            .output()
+            .with_context(|| format!("Couldn't launch {}", tool))?;
+        if !casr_cmd.status.success() {
+            let err = String::from_utf8_lossy(&casr_cmd.stderr)
+                .lines()
+                .collect::<Vec<&str>>()
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:200]: do you need collect?

Comment (SweetVishnya) [reply src/bin/casr-afl.rs:200]: And other places with join?

File: src/bin/casr-afl.rs [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+                .join(". ")
+        );
+    } else {
+        error!(
+            "{}",
+            String::from_utf8_lossy(&casr_cluster_d.stderr)
+                .lines()
+                .collect::<Vec<&str>>()
+                .join(" ")
+        );
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:233]: Do we need an exit here?

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: src/bin/casr-afl.rs [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+                .join(" ");
+            if err.contains("Program terminated (no crash)") {
+                warn!("{}: no crash on input {}", tool, crash.path.display());
+            } else {
+                error!("{} for input: {}", err, crash.path.display());
+            }
+        }
+    }
+
+    // Deduplicate reports.
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:210]: Call dedup when number of reports > 1. If there is no reports print error.

File: src/bin/casr-afl.rs [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+        error!(
+            "{}",
+            String::from_utf8_lossy(&casr_cluster_d.stderr)
+                .lines()
+                .collect::<Vec<&str>>()
+                .join(" ")
+        );
+    }
+
+    if !matches.is_present("no-cluster") {
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:236]: cluster when number of reports after dedup > 1

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: src/bin/casr-afl.rs [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+    pub path: PathBuf,
+    /// Target command line args.
+    pub target_args: Vec<String>,
+    /// Stdin.
+    pub is_stdin: bool,
+    /// ASAN.
+    pub is_asan: bool,
+}
+
+fn main() -> error::Result<()> {
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:30]: We need to remove custom error types like Casr-san/Casr-cluster, etc. Now all anyhow errors cast to Casr-cluster error. This leads to wrong message:

Error: Cluster("No such file or directory (os error 2). Couldn't launch casr-san. ")

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: src/bin/casr-afl.rs [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+                crash_info.path = crash_path.to_path_buf();
+                crashes.push(crash_info.clone());
+            }
+        }
+    }
+
+    // Generate CASR reports.
+    info!("Generating CASR reports...");
+    let output_dir = Path::new(matches.value_of("output").unwrap());
+    if !output_dir.exists() {
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:155]: Test that output directory is empty?

File: src/bin/casr-afl.rs (hunk 7 above)
Comment (anfedotoff) [reply src/bin/casr-afl.rs:200]: Maybe for error printing we don't need. I think I made Fix for casr-cluster to print results on single line.

File: src/bin/casr.rs [hunk 13]
```diff
@@ -44,7 +44,7 @@ fn main() -> Result<()> {
             .long("mode")
             .takes_value(true)
             .value_name("MODE")
-            .possible_values(&["online", "offline"])
```
[the remark this answers is not in the document]
Comment (anfedotoff) [reply src/bin/casr.rs:47]: Here there are also fixes for casr-afl.rs I fixed clippy on master also.

File: src/bin/casr-afl.rs (hunk 3 above)
Comment (anfedotoff) [reply src/bin/casr-afl.rs:164]: We don't know number of crashes before we start generating reports. Deduplication phase prints reports `before`. Do we need to count crashes while attempting to generate report?

File: src/bin/casr-afl.rs (hunk 2 above)
Comment (anfedotoff) [reply src/bin/casr-afl.rs:119]: Maybe later this is one place only. I think, it's OK for now.

File: src/bin/casr-afl.rs (hunk 3 above)
Comment (SweetVishnya) [reply src/bin/casr-afl.rs:164]: Ok, no need

File: docs/usage.md [hunk 14]
```diff
@@ -217,16 +217,74 @@ Triage crashes found by AFL++
...
-**NOTE:** `casr-gdb` and `casr-san` should be in PATH to make `casr-afl` works.
+**NOTE:** `casr-gdb` and `casr-san` should be in PATH to make `casr-afl` work.
 
 Example (Ubuntu 20.04+):
 
     $ cp tests/casr_tests/bin/load_afl /tmp/load_afl
     $ cp tests/casr_tests/bin/load_sydr /tmp/load_sydr
     $ casr-afl -i tests/casr_tests/bin/afl-out-xlnt -o tests/tmp_tests_casr/casr_afl_out
+
+    $ tree casr_afl_out
```
Comment (SweetVishnya) [inline docs/usage.md:235]: the path is wrong

File: src/bin/casr-afl.rs [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+        let mut args: Vec<String> = vec!["-o".to_string()];
+        if crash.is_asan {
+            args.push(format!(
+                "{}.casrep",
+                output_dir.join(crash.path.file_name().unwrap()).display()
+            ));
+        } else {
+            args.push(format!(
+                "{}.gdb.casrep",
+                output_dir.join(crash.path.file_name().unwrap()).display()
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:175]: repeats, maybe place to variable?

File: src/bin/casr-afl.rs [hunk 16]
```diff
@@ -171,34 +175,30 @@ fn main() -> error::Result<()> {
...
-                args[at_index] = input;
-            }
         }
         let tool = if crash.is_asan {
             "casr-san"
         } else {
             "casr-gdb"
         };
-        debug!("{} {}", tool, args.join(" "));
+        debug!("{} {:?}", tool, args);
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:195]: Command structure is printable

File: src/bin/casr-afl.rs [hunk 17]
```diff
@@ -89,6 +88,15 @@ fn main() -> error::Result<()> {
         TerminalMode::Mixed,
     );
 
+    let output_dir = Path::new(matches.value_of("output").unwrap());
+    if !output_dir.exists() {
+        fs::create_dir(output_dir).with_context(|| {
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:93]: create_dir_all?

File: src/bin/casr-afl.rs [hunk 18]
```diff
@@ -89,6 +88,15 @@ fn main() -> error::Result<()> {
...
+            format!("Couldn't create output directory {}", output_dir.display())
+        })?;
+    } else if output_dir.read_dir()?.next().is_some() {
+        error!("Output directory is not empty.");
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs]: Maybe bail?

File: src/error.rs [hunk 19]
```diff
@@ -11,8 +11,6 @@ pub enum Error {
     IO(io::Error),
     /// Casr-cluster error
     Cluster(String),
```
Comment (SweetVishnya) [inline src/error.rs:13]: Why do we need this?

File: src/bin/casr-cluster.rs [hunk 20]
```diff
@@ -500,8 +481,10 @@ fn main() -> error::Result<()> {
             .map(Path::new)
             .collect();
         let (before, after) = dedup(paths[0], paths.get(1).map(|x| x.to_path_buf()))?;
-        println!("Number of reports before deduplication: {}", before);
-        println!("Number of reports after deduplication: {}", after);
+        println!(
```
Comment (SweetVishnya) [inline src/bin/casr-cluster.rs:484]: Better revert and join strings in casr-afl

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: src/bin/casr-afl.rs [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+            "casr-san"
+        } else {
+            "casr-gdb"
+        };
+        let mut casr_cmd = Command::new(tool);
+        casr_cmd.args(&args);
+        debug!("{:?}", casr_cmd);
+        let casr_output = casr_cmd
+            .output()
+            .with_context(|| format!("Couldn't launch {} {:?}", tool, args))?;
```
Comment (SweetVishnya) [inline src/bin/casr-afl.rs:195]: print casr_cmd

**Review by SweetVishnya — APPROVED** [review]

**Pull request merged**